validator/project_validator.go

```diff
@@ -1850,178 +1804,68 @@ func validateTaskSyncCommands(p *model.Project, runLong bool) ValidationErrors {
-	}
+		if dependentTaskUnit.RunsOnPatches() && (!dependedOnTaskUnit.RunsOnPatches() || dep.PatchOptional) {
+			return false
+		}
+		if dependentTaskUnit.RunsOnNonPatches() && !dependedOnTaskUnit.RunsOnNonPatches() {
+			return false
+		}
+		if dependentTaskUnit.RunsOnGitTag() && !dependedOnTaskUnit.RunsOnGitTag() {
+			return false
+		}
 
-	taskUnit := tvToTaskUnit[current]
-	// Even if current depends on target according to the dependency graph, if
-	// the current task will not run in the same cases as the source (e.g. the
-	// source task runs on patches but current task does not, or if the current task
-	// is only available to git tag builds), the dependency is
-	// not reachable from this branch.
-	if depReqs.requireOnPatches && (taskUnit.SkipOnPatchBuild() || taskUnit.SkipOnNonGitTagBuild()) {
-		return false, nil
-	}
-	if depReqs.requireOnNonPatches && (taskUnit.SkipOnNonPatchBuild() || taskUnit.SkipOnNonGitTagBuild()) {
-		return false, nil
-	}
-	if depReqs.requireOnGitTag && (taskUnit.SkipOnNonPatchBuild() || taskUnit.SkipOnGitTagBuild()) {
-		return false, nil
-	}
+		if statuses != nil && dependedOn == dependedOnNode {
+			return utility.StringSliceContains(statuses, dep.Status)
+		}
 
-	if current == target {
-		return depReqs.lastDepNeedsSuccess, nil
+		return true
 	}
 
-	visited[current] = true
-
-	depsToNodes := dependenciesForTaskUnit(current, taskUnit, allNodes)
-	for dep, depNodes := range depsToNodes {
-		// If the task must run on patches but this dependency is optional on
-		// patches, we cannot traverse this dependency branch.
-		if depReqs.requireOnPatches && dep.PatchOptional {
-			continue
+	if found := g.DepthFirstSearch(dependentNode, dependedOnNode, traversal); !found {
+		dependentBVTask := tvTaskUnitMap[dependentTask]
+		errMsg := "task '%s' in build variant '%s' must depend on" +
+			" task '%s' in build variant '%s' completing"
+		if dependentBVTask.RunsOnPatches() && dependentBVTask.RunsOnNonPatches() {
+			errMsg += " for both patches and non-patches"
+		} else if dependentBVTask.RunsOnPatches() {
+			errMsg += " for patches"
+		} else if dependentBVTask.RunsOnNonPatches() {
+			errMsg += " for non-patches"
+		} else if dependentBVTask.RunsOnGitTag() {
+			errMsg += " for git-tag builds"
 		}
-		depReqs.lastDepNeedsSuccess = dep.Status == "" || dep.Status == evergreen.TaskSucceeded
+		errMsg = fmt.Sprintf(errMsg, dependentTask.TaskName, dependentTask.Variant, dependedOnTask.TaskName, dependedOnTask.Variant)
 
-		for _, depNode := range depNodes {
-			reachable, err := dependencyMustRun(target, depNode, depReqs, allNodes, visited, tvToTaskUnit)
-			if err != nil {
-				return false, errors.Wrap(err, "dependency graph has problems")
-			}
-			if reachable {
-				return true, nil
-			}
+		if statuses != nil {
+			errMsg = fmt.Sprintf(errMsg+" with status in [%s]", strings.Join(statuses, ", "))
```

Comment:
Could we instead do `fmt.Sprintf("%s with status in [%s]", errMsg, strings.Join...)` ?

Reply:
done